modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Assets.scala

```diff
@@ -0,0 +1,289 @@
+     * @param stop global stop signal from [[Environment]]
+     * @param assets all assets that have to be tracked
+     */
+    def make[F[_]: ConcurrentEffect: Timer: ContextShift](
+      blocker: Blocker,
+      stop: Ref[F, Boolean],
+      assets: List[Asset]
+    ): Resource[F, State[F]] =
+      for {
+        clients <- Clients.make[F](blocker, assets.map(_._1))
+        map <- Resource.liftF(build[F](blocker, clients, assets.filterNot(asset => asset == TestPair)))
+        files <- Resource.liftF(Ref.of[F, Map[URI, Hash]](map))
+      } yield State(files, stop, clients)
+
+    def build[F[_]: Concurrent: Timer: ContextShift](
+      blocker: Blocker,
+      clients: Clients[F],
+      assets: List[Asset]
+    ): F[Map[URI, Hash]] =
+      buildFromLocal(blocker, assets)
+        .flatMap { hashes =>
+          hashes.traverse {
+            case (uri, path, Some(hash)) =>
+              Logger[F].info(s"Asset from $uri is found on local system at $path").as(uri -> hash)
+            case (uri, path, None) =>
+              downloadAndHash[F](clients, blocker, uri, Paths.get(path)).map(hash => uri -> hash)
+          }
+        }
+        .map(_.toMap)
+
+    def buildFromLocal[F[_]: Sync: ContextShift](blocker: Blocker, assets: List[Asset]): F[List[(URI, String, Option[Hash])]] =
+      assets.traverse { case (uri, path) => local[F](blocker, path).map(hash => (uri, path, hash)) }
+
+    /** Check if file already exists */
+    def local[F[_]: Sync: ContextShift](blocker: Blocker, path: String): F[Option[Hash]] = {
+      val fpath = Paths.get(path)
+      exists(blocker, fpath).ifM(
+        readAll(fpath, blocker, 1024).through(md5).compile.to(Array).map(b => Hash(b).some),
+        Sync[F].pure(none)
+      )
+    }
+  }
+
+  final case class Hash private (s: String) extends AnyVal
+
+  object Hash {
+    def apply(bytes: Array[Byte]): Hash = {
+      val bi = new java.math.BigInteger(1, bytes)
+      Hash(String.format("%0" + (bytes.length << 1) + "x", bi))
+    }
+  }
+
+  /** Pair of a tracked `URI` and destination path on local FS (`java.nio.file.Path` is not serializable) */
+  type Asset = (URI, String)
+
+  /** Initialise the [[updateStream]] with all necessary resources if refresh period is specified */
+  def run[F[_]: ConcurrentEffect: ContextShift: Timer: Parallel](env: Environment[F]): Stream[F, Unit] =
+    env.assetsUpdatePeriod match {
+      case Some(duration) =>
+        val init = for {
+          curDir <- getCurDir
+          _ <- Logger[F].info(s"Initializing assets refresh stream in $curDir")
+          assets <- env.enrichments.get.map(_.configs.flatMap(_.filesToCache))
+          _ <- download(env.blocker, env.assetsState, curDir, assets)
+          _ <- env.pauseEnrich.set(false)
+        } yield updateStream[F](env.blocker, env.assetsState, env.enrichments, curDir, duration, assets)
+        Stream.eval(init).flatten
+      case None =>
+        Stream.empty.covary[F]
```

Comment:
Don't we want to download the assets once for good even if update period is not specified ?

Reply:
That's actually a very good point. Right now we do this during [`Environment` creation](https://github.com/snowplow/enrich/blob/cc5d6b1df2dbdaada9e27560849db46e604f1540/modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Environment.scala#L127), but maybe this is a much better place.

Reply:
Argh. Just started to refactor it and remembered why I've put initial assets download into `Environment` initialization and not here. It's because we need to have all enrichment assets at the moment when enrichment configs are parsed and initialized, because SCE requires file to be there at the initialization point.

So we cannot do that unfortunately. I'll add a comment.